// @framer/rmarinac89.pulse/code/Pulse.tsx
import * as React from "react"
import { Frame, addPropertyControls, ControlType, Color } from "framer"

export function Pulse(props) {
    const { tint, duration } = props

    const marker = {
        yoyo: Infinity,
        ease: "easeInOut",
        duration: duration,
    }

    const wave = {
        loop: Infinity,
        ease: "easeOut",
        duration: duration,
        repeatDelay: duration * 1.13,
    }

    const color = Color(tint)
    const transparentColor = Color.alpha(color, 0.5)
    const radius = "100%"

    return (
        <Frame size={"100%"} borderRadius={radius} background={"transparent"}>
            <Frame
                borderRadius={radius}
                size={"100%"}
                center
                transition={wave}
                animate={{ scale: 1.35, opacity: 0 }}
                background={transparentColor}
            ></Frame>
            <Frame
                borderRadius={radius}
                transition={marker}
                animate={{ scale: 1.25, opacity: 1 }}
                size={"60%"}
                center
                background={tint}
            ></Frame>
        </Frame>
    )
}

Pulse.defaultProps = {
    height: 28,
    width: 28,
}

addPropertyControls(Pulse, {
    tint: {
        type: ControlType.Color,
        defaultValue: "rgb(0,121,255)",
    },
    duration: {
        type: ControlType.Number,
        defaultValue: 1.3,
        min: 0.1,
        max: 5,
        unit: "s",
        step: 0.1,
        displayStepper: true,
    },
})
